### ISSUE TITLE
[BUG] Using "compression" option crashes "imwrite"

### ISSUE THREAD

**Issue by ggirelli** [cgohlke/tifffile#20]
Great package, have been using it for a long time! I recently upgraded to `2020.7.24` (in Python 3.8.2) and now, when running something like
```python
import tifffile
tifffile.imwrite("test.tiff", np.zeros((10, 10, 10)), shape=(10, 10, 10), dtype=np.int8, compress=True)
```
I get the following error:
```python
  File "<stdin>", line 1, in <module>
  File "<string>", line 1, in <module>
  File "/home/gire/.local/lib/python3.8/site-packages/tifffile/tifffile.py", line 778, in imwrite
    return tif.save(data, shape, dtype, **kwargs)
  File "/home/gire/.local/lib/python3.8/site-packages/tifffile/tifffile.py", line 2096, in save
    page = next(dataiter).reshape(storedshape[1:])
TypeError: 'numpy.ndarray' object is not an iterator
```
Everything works perfectly when the `shape` argument is not provided, e.g., `tifffile.imwrite("test.tiff", np.zeros((10, 10, 10)), dtype=np.int8, compress=True)`.

I understand (might be wrong) that `shape` and `dtype` should be provided only when `data=None`, but it I think it would be better to get a more readable error saying not to use them with a `data != None` or something similar :smile:

**Comment by cgohlke** [comment]
Thank you. It is a bug that saving iterable arrays fails with compression but not without. It might be better not to accept iterable arrays in this case. They defeat the purpose of writing TIFFs out-of-core. 

If the intention is to simply save a numpy array to a TIFF file, `shape` and `dtype` should not be specified. These optional parameters are for very specific use cases as documented. In other cases the parameters are ignored. Specifically, `imwrite(..., np.zeros((10, 10, 10)), dtype=np.int8)` does not create a TIFF file with SampleFormat=INT and BitsPerSample=8 as one might expect. Probably best to raise a warning if shape or dtype don't match the data array.

**Comment by cgohlke** [comment]
I think this issue has been addressed in tifffile-2020.8.13. An exception is now raised if the dtype argument does not match the dtype of the iterable data:
```Python
>>> import numpy as np
>>> import tifffile

>>> tifffile.imwrite("test.tiff", np.zeros((10, 10, 10)), shape=(10, 10, 10), dtype=np.int8, compress=6)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "...\tifffile\tifffile.py", line 781, in imwrite
    return tif.save(data, shape, dtype, **kwargs)
  File "...\tifffile\tifffile.py", line 2133, in save
    raise ValueError('dtype of iterable does not match dtype')
ValueError: dtype of iterable does not match dtype

>>> tifffile.imwrite("test.tiff", np.zeros((10, 10, 10)), shape=(10, 10, 10), dtype=np.int8)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "...\tifffile\tifffile.py", line 781, in imwrite
    return tif.save(data, shape, dtype, **kwargs)
  File "...\tifffile\tifffile.py", line 2093, in save
    raise ValueError(
ValueError: dtype of iterable does not match dtype

>>> tifffile.imwrite("test.tiff", np.zeros((10, 10, 10)), shape=(10, 10, 10), dtype=np.float64, compress=6)

>>> tifffile.imwrite("test.tiff", np.zeros((10, 10, 10)), shape=(10, 10, 10), dtype=np.float64)

>>> tifffile.imwrite("test.tiff", np.zeros((10, 10, 10), np.int16), shape=(10, 10, 10), dtype=np.int16, compress=6)

>>> tifffile.imwrite("test.tiff", np.zeros((10, 10, 10), np.int16), shape=(10, 10, 10), dtype=np.int16)

>>> tifffile.imwrite("test.tiff", np.zeros((10, 10, 10), np.int16), compress=6)

>>> tifffile.imwrite("test.tiff", np.zeros((10, 10, 10), np.int16))
```
